README.md
# NAME

Fey - Better SQL Generation Through Perl

# VERSION

version 0.44

# SYNOPSIS

    use Fey::Literal::Function;
    use Fey::Placeholder;
    use Fey::Schema;
    use Fey::SQL;


    my $schema = hand_waving();

    my $user  = $schema->table('User');
    my $group = $schema->table('Group')

    my $select = Fey::SQL->new_select();

    my $func = Fey::Literal::Function->new( 'LCASE', $user->column('username') );

    $select->select( $user->columns( 'user_id', 'username' ) )
           ->from( $user, $group )
           ->where( $group->group_id, 'IN', 1, 2, 3 )
           ->and  ( $func, 'LIKE', 'smith%' );

    print $select->sql($dbh);

# DESCRIPTION

The `Fey` distribution contains a set of modules for representing the
components of a DBMS schema, and for dynamically generating SQL
queries based on that schema.

# USAGE

Loading this module does nothing. It's just here to provide docs and a
version number for the distro.

You'll want to take a look at [Fey::Schema](https://metacpan.org/pod/Fey::Schema), [Fey::Table](https://metacpan.org/pod/Fey::Table), and other
modules in the distro for more details.

# WHAT IS Fey?

The goal of the core `Fey` distro is to provide a simple, flexible
way of _dynamically_ generating complex SQL queries in Perl. Other
packages build on top of this functionality to create a complete ORM
(`Fey::ORM`).

# GETTING STARTED

If you're interested in an ORM, take a look at the `Fey::ORM` distro.

To generate SQL with Fey, you first need to create a set of objects
representing the tables and foreign keys in your schema. The simplest
way to do this is to use the `Fey-Loader` distro, which will connect
to an existing schema and generate a set of objects for you.

Alternatively, you can create these objects via Fey's API. You would
first create a [Fey::Schema](https://metacpan.org/pod/Fey::Schema) object. This object will hold all of
your tables and foreign keys. If you want to create your schema this
way, you should start with the [Fey::Schema](https://metacpan.org/pod/Fey::Schema), [Fey::Table](https://metacpan.org/pod/Fey::Table), and
[Fey::FK](https://metacpan.org/pod/Fey::FK) APIs. You'll also want to use the [Fey::Column](https://metacpan.org/pod/Fey::Column) API.

Once you have a schema, you can generate SQL using [Fey::SQL](https://metacpan.org/pod/Fey::SQL), or a
DBMS-specific subclass of [Fey::SQL](https://metacpan.org/pod/Fey::SQL).

# THE CORE Fey DISTRO

The emphasis in the core Fey distro is on dynamic queries,
particularly on the tables/columns/etc involved in the query, not just
the bound parameters.

This is _not_ what I mean by a dynamic query ...

    SELECT user_id FROM User where username = ?

While this is dynamic in the sense that the username is parameterized
and may change on each invocation, it is still easily handled by a
phrasebook class. If that is all you need, I suggest checking out any
of `Class::Phrasebook::SQL`, `Data::Phrasebook`, or `SQL::Library`
on CPAN.

Imagine that we have a database with a User table and a Message table,
where each message has a user who is that message's creator. We might
want to grab all the users in the database, in which case we would do
a simple `SELECT` against the User table ...

    SELECT * FROM User

But maybe we want to get all the users who have created a message in
the last week:

    SELECT User.*
      FROM User JOIN Message
           USING (user_id)
     WHERE Message.creation_date >= ?

The resultset for our query is still the same (0+ users) but the
constraints of the query are more complex. Now imagine another dozen
or so permutations on how we search for users. This is what I mean by
"dynamically" generating queries.

# RATIONALE

You probably don't need to read this if you just wanted to know how to
use Fey.

## Why Not Use a Phrasebook?

Let's assume we have a simple User table with the following columns:

    username
    state
    first_name
    last_name
    access_level

Limiting ourselves to queries of equality ("username = ?", "state =
?"), we would still need 32 (1 + 5 + 10 + 10 + 5 + 1) entries to
handle all the possible combinations of columns. Now imagine adding in
variants like allowing for wildcard searches using LIKE or regexes, or
more complex variants involving an "OR" in a subclause.

This gets even more complicated if you start adding in joins, outer
joins, and so on. It's plain to see that a phrasebook gets too large
to be usable at this point. You'd probably have to write a program
just to generate the phrasebook and keep it up to date!

## Why Not String Manipulation?

The next idea that might come to mind is to dump the phrasebook in
favor of string manipulation. This is simple enough at first, but
quickly gets ugly. Handling all of the possible options correctly
requires lots of fiddly code that has to concatenate bits of SQL in
the correct order, taking into account where to put in commas,
`WHERE` vs `AND`, and so on and so forth. I've been there, and trust
me, it's madness.

## The Solution

The core Fey modules provide a solution to the dynamic SQL
problem. Using Fey, you can specify queries in the form of _Perl
methods and objects_. Fey provides a set of objects to represent the
parts of a schema, specifically tables, columns, and foreign
keys. Using these objects along with [Fey::SQL](https://metacpan.org/pod/Fey::SQL), you can easily
generate very complex queries.

This core distro is also intended to be the foundation for building
higher-level tools like an ORM. See `Fey::ORM` for just such a thing.

# HISTORY AND GOALS

This module comes from my experience writing and using Alzabo. Alzabo
does everything this module does, and a lot more. The fact that Alzabo
does so many things has become a fairly problematic in its
maintenance, and Alzabo was over 6 years old at the time this project
was begun (August of 2006).

## Goals

Rather than coming up with a very smart solution that allows us to use
80% of a DBMS's functionality, I'd rather come up with a solution
that's dumber but supports all (or at least 99%) of the DBMS's
features. It's easy to add smarts on top of a dumb layer, but it can
be terribly hard to add that last 20% once you've got something really
smart.

The goals for Fey, based on my experience with Alzabo, are the
following:

- Provide a simple way to generate queries dynamically. I really like
how this works with Alzabo conceptually, but Alzabo is not as flexible
as I'd like and it's "build a data structure" approach to query
building can become very cumbersome.

    Rather than complex data structures, with Fey you call methods on a
    `Fey::SQL` object to build up a query. This turns out to be simpler
    to work with.

    Fey, unlike Alzabo, can be used to generate multi-row updates and
    deletes, and it supports sub-selects, unions, etc. and all that other
    good stuff.

- Fey supports complex query creation with less fiddliness than
Alzabo. This means that the class to represent queries is a little
smarter and more flexible about the order in which bits are added.

    For example, in using Alzabo I often came across cases where I wanted
    to add a table to a query's join _if it hasn't already been
    added_. With Alzabo, there's no nice clean way to do this. Simply
    adding the table to the join parameter twice will cause an error. It
    would be nice to simply be able to do this

        $select->join( $foo_table => $bar_table );

    and have it do the right thing if that join already exists (where the
    right thing is just do nothing). `Fey::SQL` does exactly that.

- Provide the core for an RDBMS-OO mapper similar to a combination of
`Alzabo::Runtime::Row` and `Class::AlzaboWrapper`.

    At the same time, query generation and the ORM are decoupled. You can
    use [Fey::SQL](https://metacpan.org/pod/Fey::SQL) to generate queries without having to every use the
    `Fey::ORM` ORM.

- Be declarative like Moose. In particular, the `Fey::ORM` ORM is as
declarative as possible, and aims to emulate Moose's declarative sugar
style where possible.
- Leverage the API user's SQL knowledge. Building up queries with Fey
looks enough like SQL that you shouldn't have to think _too_ hard
about it. This means join support is baked in at a core level, as are
subselects and ideally anything else you can do in SQL.

## Problems with Alzabo

Here are some of the problems I've had with Alzabo over the years
which inspired me to create Fey ...

- Adding support for a new DBMS to Alzabo is a lot of work, so it only
supports MySQL and Postgres. Alzabo tries to be really smart about
preventing users from shooting themselves in the foot, and requires a
lot of DBMS-specific code to achieve this.

    In retrospect, being a lot dumber and allowing for foot-shooting makes
    supporting a new DBMS much easier. People generally know how their
    DBMS works, and if they generate an invalid query or table name, it
    will throw an error.

    For example, while Fey can accommodate per-DBMS query (sub)classes, it does
    not include any by default, and is capable of supporting many DBMS-specific
    features without per-DBMS classes.

- Alzabo has too much DBMS-specific knowledge. If you want to use a SQL
function in a query, you have to import a corresponding Perl function
from the appropriate `Alzabo::SQLMaker`, which limits you to what's
already defined, or forces you to go through a cumbersome API to
define a new SQL function for use in your Perl code.

    By contrast, Fey has simple generic support for arbitrary functions
    via the `Fey::Literal::Function` class. If you need more flexibility
    you can use the `Fey::Literal::Term` subclass to generate an
    arbitrary snippet to insert into your SQL.

    A related problem is that Alzabo doesn't support multiple versions of
    a DBMS very well. Either it doesn't work with an older version at all,
    or it doesn't support some enhanced capability of a newer version. It
    mostly supports whatever version I happened to be using when I wrote a
    specific piece of functionality.

- There are now free GUI design tools for specific databases that do a
better job of supporting the database in question than Alzabo ever
has.
- Alzabo separates its classes into Create (for generation of DDL) and
Runtime (for DML) subclasses, which might have been worth the memory
savings six years ago, but just makes for an extra hassle now.
- When I originally developed Alzabo, I included a feature for
generating high-level application object classes which subclass the
Alzabo classes and add "business logic" methods. This is what is
provided by `Alzabo::MethodMaker`.

    Nowadays, I prefer to have my business logic classes simply use the
    Alzabo classes. In other words, I now prefer "has-a" and "uses-a"
    versus "is-a" object design for this case.

    Method auto-generation based on a specific schema can be quite handy,
    but it should be done in the domain-specific application classes, not
    as a subclass of the core functionality.

- Storing schemas in an Alzabo-specific format is problematic for many
reasons. It's simpler to simply get the schema definition from an
existing schema, or to allow users to define it in code.
- Alzabo's referential integrity checking code was really cool back when
I mostly used MySQL with MYISAM tables. Now it's just a maintenance
burden and a barrier for new features.
- I didn't catch the testing bug until quite a while after I'd started
working on Alzabo. Alzabo's test suite is nasty. Fey is built with
testability in mind, and high test coverage is part of my ongoing
goals for the project.
- Alzabo does too many things, which makes it hard to explain and
document.

# WHY IS IT NAMED Fey?

When I first started working on Fey, it was named "Q". This was a nice
short name to type, but obviously unsuitable for releasing on CPAN. I
wanted a nice short name that could be used in multiple distributions,
like John Siracusa's "Rose" modules.

I was standing in the shower one day and had the following series of
thoughts leading to Fey. Reading this will may give you an unpleasant
insight into my mind. You have been warned.

- SQLy

    This module is "SQL-y", as in "related to SQL". However, this name is
    bad for a number of reasons. First, it's not clear how to pronounce
    it. It may make you think of a YACC grammar ("SQL.y"). It's a weird
    combo of upper- and lower-case letters.

- SQLy => Squall

    "SQLy" and "Squall" share a number of letters, obviously.

    Squall is a single short word, which is good. However, it's a bit
    awkward to type and has a somewhat negative meaning to me, because a
    storm can mean trouble.

- Squall => Lionheart => Faye

    Squall Lionheart is a character in Final Fantasy VIII, which IMO is
    the best Final Fantasy game before the PS2.

    The inimitable Faye Wong sang the theme song for FF VIII. I love Faye
    Wong.

- Faye => Fey

    And thus we arrive at "Fey". It's nice and short, easy to type, and
    easy to say.

    Some of its meanings are "otherworldly" or "magical". Attempting to
    combine SQL and OO in any way is certainly unnatural, and if done
    right, perhaps magical. Fey can also mean "appearing slightly
    crazy". This project is certainly that.

Yes, I'm a nerd, I know.

# BUGS

Please report any bugs or feature requests to `bug-fey@rt.cpan.org`,
or through the web interface at [http://rt.cpan.org](http://rt.cpan.org).  I will be
notified, and then you'll automatically be notified of progress on
your bug as I make changes.

Bugs may be submitted through [the RT bug tracker](http://rt.cpan.org/Public/Dist/Display.html?Name=Fey)
(or [bug-fey@rt.cpan.org](mailto:bug-fey@rt.cpan.org)).

I am also usually active on IRC as 'drolsky' on `irc://irc.perl.org`.

# DONATIONS

If you'd like to thank me for the work I've done on this module, please
consider making a "donation" to me via PayPal. I spend a lot of free time
creating free software, and would appreciate any support you'd care to offer.

Please note that **I am not suggesting that you must do this** in order for me
to continue working on this particular software. I will continue to do so,
inasmuch as I have in the past, for as long as it interests me.

Similarly, a donation made in this way will probably not make me work on this
software much more, unless I get so many donations that I can consider working
on free software full time (let's all have a chuckle at that together).

To donate, log into PayPal and send money to autarch@urth.org, or use the
button at [http://www.urth.org/~autarch/fs-donation.html](http://www.urth.org/~autarch/fs-donation.html).

# AUTHOR

Dave Rolsky &lt;autarch@urth.org>

# CONTRIBUTORS

- Aristotle Pagaltzis &lt;pagaltzis@gmx.de>
- hdp@glaive.weftsoar.net &lt;hdp@glaive.weftsoar.net>
- hdp@localhost &lt;hdp@localhost>
- hdp@rook.opensourcery.com &lt;hdp@rook.opensourcery.com>
- Oliver Charles &lt;oliver@ocharles.org.uk>

# COPYRIGHT AND LICENCE

This software is Copyright (c) 2011 - 2016 by Dave Rolsky.

This is free software, licensed under:

    The Artistic License 2.0 (GPL Compatible)


Changes
{{$NEXT}}

0.43    2015-06-15

- I meant to require MooseX::Params::Validate 0.21 or greater in the 0.41
  release but somehow only required 0.08. This also increases several other
  prereq versions. This distro now requires Moose 2.1200+,
  MooseX::Role::Parameterized 1.00+, and MooseX::StrictConstructor 0.13+.


0.42    2015-06-14

- Fixed the repo info in the distro metadata.


0.41    2015-06-14

- Shut up warning from tests with newer Perls.

- Fix test failures when using MooseX::Params::Validate 0.21 or
  newer. Reported by Andreas Koenig. RT #102342.


0.40    2011-02-16

- Bind params in a subselect used in the SELECT clause of a query were not
  returned by the bind_params() method.


0.39    2011-01-07

- The value used for setting a column in an UPDATE or INSERT can now be any
  object which does the Fey::Role::SQL::ReturnsData role, which includes
  Fey::SQL::Select objects.


0.38    2010-12-14

- Various methods which returned lists of objects, such as Schema->tables()
  and Table->columns() without arguments, did not return the objects in any
  specific order. Now the objects are returned in the same order as they were
  added originally. This also applies to lists of columns in both candidate
  and foreign keys. RT #63811. Reported by John Dasch.


0.37    2010-09-25

- Removing a candidate key left an undef value in the internal candidate key
  array. I'm not sure if this could actually show up in normal use. Note that
  this fix also avoids an error that will show up in future versions of Moose
  when a native delegation method is called incorrectly.


0.36    2010-08-27

- Avoid warnings from the latest versions of Moose (1.10).


0.35    2010-06-15

- You can now include a WHERE clause as part of an inner join, like 

  FROM Part
       JOIN MachinePart
       ON Part.part_id = MachinePart.part_id
       AND MachinePart.machine_id = ?

  Implemented by Dieter Pearcey.

- Aliasable objects now have an alias method which sets the alias and returns
  the object itself, for chaining:

  my $function = Fey::Literal::Function->new( ... )->alias('bar');

  Note that this may be deprecated in a future release where we start allowing
  aliases to be set in the constructor.

  Implemented by Dieter Pearcey.

- We now use MooseX::Types internally. Patch by Oliver Charles.

- License is now Artistic 2.0, instead of "same as Perl itself". Artistic 2.0
  allows you to redistribute under the GPL (any version) if you like.


0.34    2009-10-25

- Add a can_have_alias() attribute to Fey::Term::Literal. Sometimes you want
  make sure a term is used exactly as-is, without any alias being added.

- Fey::Literal is no longer the parent class for all literal classes. Instead,
  it is simply a factory. All the literals now do the Fey::Role::IsLiteral
  role.


0.33    2009-10-07

- Put a $VERSION in every .pm file matching Fey.pm

- The $table->aliased_columns() method can now be called with just a prefix,
  in which case it returns aliases for all the columns in the table. Requested
  by Aristotle Pagaltzis. RT #40857.

- Fixed some tests in t/SQL/Set.t that changed results when modules were
  loaded in a different order.


0.32    2009-09-10

- Shut up some warnings from the next Moose (0.90).

- Improved docs on WHERE clauses, especially on what can be compared (columns,
  literals, etc.). See Fey::SQL for details. Requested by Martin. RT #48677.

- Tightened up checking of WHERE clause components. Previously an unblessed
  reference would've ended up getting stringified, now it will be rejected.


0.31    2009-08-18

- Added a new (empty) role, Fey::Role::SQL::ReturnsData. Use this as a
  constraint any place you might be checking if an object
  ->isa('Fey::SQL::Select'). This role is applied to any SQL query that
  returns data, such as SELECTs, UNIONs, etc.

- Added Fey::SQL::Select->distinct_on($column).


0.30    2009-08-12

- Allow the first argument to ->limit to be undef, so you can do
  ->limit( undef, 20 ), which produces SQL like "OFFSET 20".


0.29    2009-08-12

- Allow subselects on the right hand side of a where clause comparison using
  =, !=, or <>.


0.28    2009-05-14

- Fixed the tests to pass with newer Moose. No code changes.


0.27    2009-04-18

- All SQL objects now implement a clone() method.


0.26    2009-04-04

- Some uses of Fey::SQL::Fragment::SubSelect were still lingering in
  the code, so a fresh install failed. Reported by Chad Granum.


0.25    2009-04-01

- Major internals changes to how aliasable things are created and
  used. Added two new roles, Fey::Role::MakesAliasObjects, and
  Fey::Role::HasAliasName. All literals can now be aliased (including
  Terms). A subselect is now just a Fey::SQL::Select object
  internally, which makes the internals much cleaner. This also means
  that a Fey::SQL::Select has an ->alias_name. Implemented by Dieter
  Pearcey.

- Added support for "NULLS FIRST/LAST" in ORDER BY
  clauses. Implemented by Erik Rijkers. RT #44656.


0.24    2009-03-26

- Support for UNION, INTERSECT, and EXCEPT queries. Implemented by
  Dieter Pearcey.


0.23    2009-03-23

- Fey::SQL::Insert did not make sure that the order of the
  bind_params() it returned actually matched the order of the columns
  in the "INSERT ... INTO" clause. It accidentally worked sometimes
  but was fundamentally broken. This only mattered if you were using
  the auto_placeholders feature, so it did not affect
  Fey::ORM. Reported and diagnosed by Dieter Pearcey.


0.22    2009-02-24

- A table that was passed to $sql->from($table) and also included in a
  join could show up in the FROM clause twice. Based on a patch from
  Dieter Pearcey.

- Fey::Literal::Term now accepts a list of items which it joins
  together. These items can be strings or Fey objects like columns or
  tables. This makes generating arbitrary of SQL snippets
  easier. Based on a patch from Dieter Pearcey.


0.21    2009-02-24

- If you created two Fey::Table::Alias objects with the same name

lib/Fey.pm
package Fey;

use strict;
use warnings;

our $VERSION = '0.44';

1;

# ABSTRACT: Better SQL Generation Through Perl

__END__

=head1 SYNOPSIS

  use Fey::Literal::Function;
  use Fey::Placeholder;
  use Fey::Schema;
  use Fey::SQL;


  my $schema = hand_waving();

  my $user  = $schema->table('User');
  my $group = $schema->table('Group')

  my $select = Fey::SQL->new_select();

  my $func = Fey::Literal::Function->new( 'LCASE', $user->column('username') );

  $select->select( $user->columns( 'user_id', 'username' ) )
         ->from( $user, $group )
         ->where( $group->group_id, 'IN', 1, 2, 3 )
         ->and  ( $func, 'LIKE', 'smith%' );

  print $select->sql($dbh);

=head1 DESCRIPTION

The C<Fey> distribution contains a set of modules for representing the
components of a DBMS schema, and for dynamically generating SQL
queries based on that schema.

=head1 USAGE

Loading this module does nothing. It's just here to provide docs and a
version number for the distro.

You'll want to take a look at L<Fey::Schema>, L<Fey::Table>, and other
modules in the distro for more details.

=head1 WHAT IS Fey?

The goal of the core C<Fey> distro is to provide a simple, flexible
way of I<dynamically> generating complex SQL queries in Perl. Other
packages build on top of this functionality to create a complete ORM
(C<Fey::ORM>).

=head1 GETTING STARTED

If you're interested in an ORM, take a look at the C<Fey::ORM> distro.

To generate SQL with Fey, you first need to create a set of objects
representing the tables and foreign keys in your schema. The simplest
way to do this is to use the C<Fey-Loader> distro, which will connect
to an existing schema and generate a set of objects for you.

Alternatively, you can create these objects via Fey's API. You would
first create a L<Fey::Schema> object. This object will hold all of
your tables and foreign keys. If you want to create your schema this
way, you should start with the L<Fey::Schema>, L<Fey::Table>, and
L<Fey::FK> APIs. You'll also want to use the L<Fey::Column> API.

Once you have a schema, you can generate SQL using L<Fey::SQL>, or a
DBMS-specific subclass of L<Fey::SQL>.

=head1 THE CORE Fey DISTRO

The emphasis in the core Fey distro is on dynamic queries,
particularly on the tables/columns/etc involved in the query, not just
the bound parameters.

This is I<not> what I mean by a dynamic query ...

 SELECT user_id FROM User where username = ?

While this is dynamic in the sense that the username is parameterized
and may change on each invocation, it is still easily handled by a
phrasebook class. If that is all you need, I suggest checking out any
of C<Class::Phrasebook::SQL>, C<Data::Phrasebook>, or C<SQL::Library>
on CPAN.

Imagine that we have a database with a User table and a Message table,
where each message has a user who is that message's creator. We might
want to grab all the users in the database, in which case we would do
a simple C<SELECT> against the User table ...

 SELECT * FROM User

But maybe we want to get all the users who have created a message in
the last week:

 SELECT User.*
   FROM User JOIN Message
        USING (user_id)
  WHERE Message.creation_date >= ?

The resultset for our query is still the same (0+ users) but the
constraints of the query are more complex. Now imagine another dozen
or so permutations on how we search for users. This is what I mean by
"dynamically" generating queries.

=head1 RATIONALE

You probably don't need to read this if you just wanted to know how to
use Fey.

=head2 Why Not Use a Phrasebook?

Let's assume we have a simple User table with the following columns:

 username
 state
 first_name
 last_name
 access_level

Limiting ourselves to queries of equality ("username = ?", "state =
?"), we would still need 32 (1 + 5 + 10 + 10 + 5 + 1) entries to
handle all the possible combinations of columns. Now imagine adding in
variants like allowing for wildcard searches using LIKE or regexes, or
more complex variants involving an "OR" in a subclause.

This gets even more complicated if you start adding in joins, outer
joins, and so on. It's plain to see that a phrasebook gets too large
to be usable at this point. You'd probably have to write a program
just to generate the phrasebook and keep it up to date!

=head2 Why Not String Manipulation?

The next idea that might come to mind is to dump the phrasebook in
favor of string manipulation. This is simple enough at first, but
quickly gets ugly. Handling all of the possible options correctly
requires lots of fiddly code that has to concatenate bits of SQL in
the correct order, taking into account where to put in commas,
C<WHERE> vs C<AND>, and so on and so forth. I've been there, and trust
me, it's madness.

=head2 The Solution

The core Fey modules provide a solution to the dynamic SQL
problem. Using Fey, you can specify queries in the form of I<Perl
methods and objects>. Fey provides a set of objects to represent the
parts of a schema, specifically tables, columns, and foreign
keys. Using these objects along with L<Fey::SQL>, you can easily
generate very complex queries.

This core distro is also intended to be the foundation for building
higher-level tools like an ORM. See C<Fey::ORM> for just such a thing.

=head1 HISTORY AND GOALS

This module comes from my experience writing and using Alzabo. Alzabo
does everything this module does, and a lot more. The fact that Alzabo
does so many things has become a fairly problematic in its
maintenance, and Alzabo was over 6 years old at the time this project
was begun (August of 2006).

=head2 Goals

Rather than coming up with a very smart solution that allows us to use
80% of a DBMS's functionality, I'd rather come up with a solution
that's dumber but supports all (or at least 99%) of the DBMS's
features. It's easy to add smarts on top of a dumb layer, but it can
be terribly hard to add that last 20% once you've got something really
smart.

The goals for Fey, based on my experience with Alzabo, are the
following:

=over 4

=item *

Provide a simple way to generate queries dynamically. I really like
how this works with Alzabo conceptually, but Alzabo is not as flexible
as I'd like and it's "build a data structure" approach to query
building can become very cumbersome.

Rather than complex data structures, with Fey you call methods on a
C<Fey::SQL> object to build up a query. This turns out to be simpler
to work with.

Fey, unlike Alzabo, can be used to generate multi-row updates and
deletes, and it supports sub-selects, unions, etc. and all that other
good stuff.

=item *

Fey supports complex query creation with less fiddliness than
Alzabo. This means that the class to represent queries is a little
smarter and more flexible about the order in which bits are added.

For example, in using Alzabo I often came across cases where I wanted
to add a table to a query's join I<if it hasn't already been
added>. With Alzabo, there's no nice clean way to do this. Simply
adding the table to the join parameter twice will cause an error. It
would be nice to simply be able to do this

  $select->join( $foo_table => $bar_table );

and have it do the right thing if that join already exists (where the
right thing is just do nothing). C<Fey::SQL> does exactly that.

=item *

Provide the core for an RDBMS-OO mapper similar to a combination of
C<Alzabo::Runtime::Row> and C<Class::AlzaboWrapper>.

At the same time, query generation and the ORM are decoupled. You can
use L<Fey::SQL> to generate queries without having to every use the
C<Fey::ORM> ORM.

=item *

Be declarative like Moose. In particular, the C<Fey::ORM> ORM is as
declarative as possible, and aims to emulate Moose's declarative sugar
style where possible.

=item *

Leverage the API user's SQL knowledge. Building up queries with Fey
looks enough like SQL that you shouldn't have to think I<too> hard
about it. This means join support is baked in at a core level, as are
subselects and ideally anything else you can do in SQL.

=back

=head2 Problems with Alzabo

Here are some of the problems I've had with Alzabo over the years
which inspired me to create Fey ...

=over 4

=item *

Adding support for a new DBMS to Alzabo is a lot of work, so it only
supports MySQL and Postgres. Alzabo tries to be really smart about
preventing users from shooting themselves in the foot, and requires a
lot of DBMS-specific code to achieve this.

In retrospect, being a lot dumber and allowing for foot-shooting makes
supporting a new DBMS much easier. People generally know how their
DBMS works, and if they generate an invalid query or table name, it
will throw an error.

For example, while Fey can accommodate per-DBMS query (sub)classes, it does
not include any by default, and is capable of supporting many DBMS-specific
features without per-DBMS classes.

=item *

Alzabo has too much DBMS-specific knowledge. If you want to use a SQL
function in a query, you have to import a corresponding Perl function
from the appropriate C<Alzabo::SQLMaker>, which limits you to what's
already defined, or forces you to go through a cumbersome API to
define a new SQL function for use in your Perl code.

By contrast, Fey has simple generic support for arbitrary functions
via the C<Fey::Literal::Function> class. If you need more flexibility
you can use the C<Fey::Literal::Term> subclass to generate an
arbitrary snippet to insert into your SQL.

A related problem is that Alzabo doesn't support multiple versions of
a DBMS very well. Either it doesn't work with an older version at all,
or it doesn't support some enhanced capability of a newer version. It
mostly supports whatever version I happened to be using when I wrote a
specific piece of functionality.

=item *

There are now free GUI design tools for specific databases that do a
better job of supporting the database in question than Alzabo ever
has.

=item *

Alzabo separates its classes into Create (for generation of DDL) and
Runtime (for DML) subclasses, which might have been worth the memory
savings six years ago, but just makes for an extra hassle now.

=item *

When I originally developed Alzabo, I included a feature for
generating high-level application object classes which subclass the
Alzabo classes and add "business logic" methods. This is what is
provided by C<Alzabo::MethodMaker>.

Nowadays, I prefer to have my business logic classes simply use the
Alzabo classes. In other words, I now prefer "has-a" and "uses-a"
versus "is-a" object design for this case.

Method auto-generation based on a specific schema can be quite handy,
but it should be done in the domain-specific application classes, not
as a subclass of the core functionality.

=item *

Storing schemas in an Alzabo-specific format is problematic for many
reasons. It's simpler to simply get the schema definition from an
existing schema, or to allow users to define it in code.

=item *

Alzabo's referential integrity checking code was really cool back when
I mostly used MySQL with MYISAM tables. Now it's just a maintenance
burden and a barrier for new features.

=item *

I didn't catch the testing bug until quite a while after I'd started
working on Alzabo. Alzabo's test suite is nasty. Fey is built with
testability in mind, and high test coverage is part of my ongoing
goals for the project.

=item *

Alzabo does too many things, which makes it hard to explain and
document.

=back

=head1 WHY IS IT NAMED Fey?

When I first started working on Fey, it was named "Q". This was a nice
short name to type, but obviously unsuitable for releasing on CPAN. I
wanted a nice short name that could be used in multiple distributions,
like John Siracusa's "Rose" modules.

I was standing in the shower one day and had the following series of
thoughts leading to Fey. Reading this will may give you an unpleasant
insight into my mind. You have been warned.

=over 4

=item * SQLy

This module is "SQL-y", as in "related to SQL". However, this name is
bad for a number of reasons. First, it's not clear how to pronounce
it. It may make you think of a YACC grammar ("SQL.y"). It's a weird
combo of upper- and lower-case letters.

=item * SQLy => Squall

"SQLy" and "Squall" share a number of letters, obviously.

Squall is a single short word, which is good. However, it's a bit
awkward to type and has a somewhat negative meaning to me, because a
storm can mean trouble.

=item * Squall => Lionheart => Faye

Squall Lionheart is a character in Final Fantasy VIII, which IMO is
the best Final Fantasy game before the PS2.

The inimitable Faye Wong sang the theme song for FF VIII. I love Faye
Wong.

=item * Faye => Fey

And thus we arrive at "Fey". It's nice and short, easy to type, and
easy to say.

Some of its meanings are "otherworldly" or "magical". Attempting to
combine SQL and OO in any way is certainly unnatural, and if done
right, perhaps magical. Fey can also mean "appearing slightly
crazy". This project is certainly that.

=back

Yes, I'm a nerd, I know.

=head1 BUGS

Please report any bugs or feature requests to C<bug-fey@rt.cpan.org>,
or through the web interface at L<http://rt.cpan.org>.  I will be
notified, and then you'll automatically be notified of progress on
your bug as I make changes.

=cut
